keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,560 @@
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
+    // +optional
+    FileRef *FileEnvSource `json:"fileRef,omitempty" protobuf:"bytes,3,opt,name=fileRef"`
+    ...
+}
+
+type FileEnvSource struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // Specify whether the file must exist.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
+}
+```
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file.
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
+### Env File
+
+The full specification of an env file:
+
+1. **File Format**: The environment variable (env) file must adhere to valid YAML syntax to ensure correct parsing.
+
+2. **Variable Naming**:
+    
+    a. **Standard**: Environment variable names should be composed of alphanumeric characters (letters and numbers), underscores, dots, or hyphens. The first character cannot be a digit.
+
+    b. **Relaxed**: If the Kubernetes feature gate `RelaxedEnvironmentVariableValidation` is enabled, the naming restriction is less stringent, allowing any printable ASCII character except the equals sign (=).
+
+3. **Duplicate Names**: If an environment variable is defined multiple times in the file, the last occurrence takes precedence and overrides any previous values.
+
+4. **Size Limit**: The maximum allowed size for the env file is 1 MiB.
+
+5. **File Location**: The env file must be placed within the `emptyDir` volume associated with the pod. If it is not found in the correct location, the Kubernetes API server will reject the pod creation request.
+
+6. **Container Behavior**:
+    
+    a. **Startup**: At container startup, the kubelet (the Kubernetes node agent) will parse the env file from the emptyDir volume and inject the defined variables into the container's environment.
```

Comment:
This could lead to a race condition if another container is updating the file at the same time a container is starting. More specifically, we need to be really careful when reading the file to avoid symlink shenanigans  (see https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/volume/util/subpath/subpath_linux.go#L530)

Reply:
+1

Reply:
To avoid the race condition, we can restrict mounting the `emptyDir` volume (containing the env file) only in one `initContainer`.